swift/README.md

```diff
@@ -1,372 +1,1100 @@
+Prefer compact code and let the compiler infer the type for constants or variables of single instances. Type inference is also appropriate for small (non-empty) arrays and dictionaries. When required, specify the specific type such as `CGFloat` or `Int16`.
 
+**Preferred:**
 ```swift
-XCTAssertEqual(actualValue, expectedValue)
+let message = "Click the button"
+let currentBounds = computeViewBounds()
+var names = ["Mic", "Sam", "Christine"]
+let maximumWidth: CGFloat = 106.5
+```
+
+**Not Preferred:**
+```swift
+let message: String = "Click the button"
+let currentBounds: CGRect = computeViewBounds()
+let names = [String]()
 ```
 
-`if`/`else`
-===========
+#### Type Annotation for Empty Arrays and Dictionaries
 
-This is the style:
+For empty arrays and dictionaries, use type annotation. (For an array or dictionary assigned to a large, multi-line literal, use type annotation.)
 
+**Preferred:**
 ```swift
-if condition {
+var names = [String]()
+var lookup = [String: Int]()
+```
 
-} else {
+**Not Preferred:**
+```swift
+var names: [String] = []
+var lookup: [String: Int] = [:]
+```
+
+**NOTE**: Following this guideline means picking descriptive names is even more important than before.
+
+### Syntactic Sugar
+
+Prefer the shortcut versions of type declarations over the full generics syntax.
+
+**Preferred:**
+```swift
+var deviceModels: [String]
+var employees: [Int: String]
+var faxNumber: Int?
+```
+
+**Not Preferred:**
+```swift
+var deviceModels: Array<String>
+var employees: Dictionary<Int, String>
+var faxNumber: Optional<Int>
+```
+
+## Functions vs Methods
+
+Free functions, which aren't attached to a class or type, should be used sparingly. When possible, prefer to use a method instead of a free function. This aids in readability and discoverability.
+
+Free functions are most appropriate when they aren't associated with any particular type or instance.
+
+**Preferred**
+```swift
+let sorted = items.mergeSort()  // easily discoverable
+rocket.launch()  // clearly acts on the model
+```
+
+**Not Preferred**
+```swift
+let sorted = mergeSort(items)  // hard to discover
+launch(&rocket)
+```
+
+**Free Function Exceptions**
+```swift
+let tuples = zip(a, b)  // feels natural as a free function (symmetry)
+let value = max(x,y,z)  // another free function that feels natural
+```
+
+## Memory Management
+
+Code (even non-production) should not create reference cycles. Analyze your object graph and prevent strong cycles with `weak` and `unowned` references. Alternatively, use value types (`struct`, `enum`) to prevent cycles altogether.
+
+Variables that are delegates should be `weak`.
+
+### Extending object lifetime
+
+Extend object lifetime using the `[weak self]` and `guard let strongSelf = self else { return }` idiom. `[weak self]` is preferred to `[unowned self]` where it is not immediately obvious that `self` outlives the closure. Explicitly extending lifetime is preferred to optional unwrapping.
```

Comment:
Yes, that makes sense.  I think when I read this I was focusing on the `[weak self]` bit.  Wonder if we could re-word this.  I think it's basically two things we need to outline:
1. Use `[weak self]` to avoid a retain cycle.
2. Within a block, create a reference to `self` immediately and use that reference throughout the block. This avoids `self` (which is weak) being released during execution of the block.